compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -148,53 +152,85 @@ struct AbstractConstBuilder<'a, 'tcx> {
 }
 
 impl<'a, 'tcx> AbstractConstBuilder<'a, 'tcx> {
-    fn new(tcx: TyCtxt<'tcx>, body: &'a mir::Body<'tcx>) -> Option<AbstractConstBuilder<'a, 'tcx>> {
-        // We only allow consts without control flow, so
-        // we check for cycles here which simplifies the
-        // rest of this implementation.
-        if body.is_cfg_cyclic() {
-            return None;
+    fn error(&mut self, span: Option<Span>, msg: &str) -> Result<!, ErrorReported> {
+        let mut err =
+            self.tcx.sess.struct_span_err(self.body.span, "overly complex generic constant");
+        if let Some(span) = span {
+            err.span_note(span, msg);
```

Comment:
Since `span` is always a sub-span of `self.body.span`, you could use `span_label` (and in case there's no span, just default the span to `self.body.span`. This should make the diagnostic less vertical and redundant